consensus/src/data_manager.rs

```diff
@@ -0,0 +1,235 @@
+                pos,
+                pos.digest(),
+                logical_time
+            );
+            if logical_time <= pos.expiration() {
+                receivers.push(
+                    self.data_reader
+                        .load()
+                        .as_ref()
+                        .unwrap() //TODO: can this be None? Need to make sure we call new_epoch() first.
+                        .get_batch(pos)
+                        .await,
+                );
+            } else {
+                debug!("QS: skipped expired pos");
+            }
+        }
+        receivers
+    }
+
+    ///Pass commit information to BatchReader and QuorumStore wrapper for their internal cleanups.
+    pub async fn notify_commit(&self, logical_time: LogicalTime, payloads: Vec<Payload>) {
+        if self.quorum_store_enabled.load(Ordering::Relaxed) {
+            self.data_reader
+                .load()
+                .as_ref()
+                .unwrap()
+                .update_certified_round(logical_time)
+                .await;
+
+            let digests: Vec<HashValue> = payloads
+                .into_iter()
+                .flat_map(|payload| match payload {
+                    Payload::DirectMempool(_) => {
+                        unreachable!()
+                    }
+                    Payload::InQuorumStore(proof_with_status) => proof_with_status.proofs,
+                    Payload::Empty => Vec::new(),
+                })
+                .map(|proof| *proof.digest())
+                .collect();
+
+            let _ = self
+                .quorum_store_wrapper_tx
+                .load()
+                .as_ref()
+                .unwrap()
+                .as_ref()
+                .clone()
+                .try_send(WrapperCommand::CleanRequest(logical_time, digests));
+        }
+    }
+
+    /// Called from consensus to pre-fetch the transaction behind the batches in the block.
+    pub async fn update_payload(&self, block: &Block) {
+        if self.quorum_store_enabled.load(Ordering::Relaxed) && block.payload().is_some() {
+            match block.payload().unwrap() {
+                Payload::InQuorumStore(proof_with_status) => {
+                    if proof_with_status.status.lock().is_none() {
+                        let receivers = self
+                            .request_transactions(
+                                proof_with_status.proofs.clone(),
+                                LogicalTime::new(block.epoch(), block.round()),
+                            )
+                            .await;
+                        proof_with_status
+                            .status
+                            .lock()
+                            .replace(DataStatus::Requested(receivers));
+                    }
+                }
+                Payload::Empty => {}
+                Payload::DirectMempool(_) => {
+                    unreachable!()
```

Comment:
That is a good point. 
@bchocho and I are considering two options, what do you think?
1. just ignore and log a warning. This is safe as all will ignore it. If a bad leader want to send as trash he can do it anyway (or send nothing which I even worst).
2. check payload enum before voting in consensus.

Since we are going to delete this code anyway after we enable quorum store, I personally prefer 1. But happy to do both changes or any better idea you might have.